receiver/stanzareceiver/e2e_test.go

```diff
@@ -0,0 +1,153 @@
+	t *testing.T
+}
+
+// ReportFatalError causes the test to be run to fail.
+func (h *testHost) ReportFatalError(err error) {
+	h.t.Fatalf("receiver reported a fatal error: %v", err)
+}
+
+var _ component.Host = (*testHost)(nil)
+
+func unmarshal(t *testing.T, pipelineYaml string) pipeline.Config {
+	pipelineCfg := pipeline.Config{}
+	require.NoError(t, yaml.Unmarshal([]byte(pipelineYaml), &pipelineCfg))
+	return pipelineCfg
+}
+
+func expectNLogs(sink *exportertest.SinkLogsExporter, expected int) func() bool {
+	return func() bool { return sink.LogRecordsCount() == expected }
+}
+
+func TestReadStaticFile(t *testing.T) {
+	t.Parallel()
+	f := NewFactory()
+	sink := &exportertest.SinkLogsExporter{}
+	params := component.ReceiverCreateParams{Logger: zaptest.NewLogger(t)}
+
+	cfg := f.CreateDefaultConfig().(*Config)
+	cfg.Pipeline = unmarshal(t, `
+- type: file_input
+  include: [testdata/simple.log]
+  start_at: beginning`)
+
+	rcvr, err := f.CreateLogsReceiver(context.Background(), params, cfg, sink)
+	require.NoError(t, err, "failed to create receiver")
+
+	require.NoError(t, rcvr.Start(context.Background(), &testHost{t: t}))
+	require.Eventually(t, expectNLogs(sink, 3), time.Second, time.Millisecond)
+	require.NoError(t, rcvr.Shutdown(context.Background()))
+}
+
+func TestReadRotatingFiles(t *testing.T) {
+	tests := []rotationTest{
+		{"CopyTruncate", true},
+		{"NoCopyTruncate", false},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, tc.Run)
+	}
+}
+
+type rotationTest struct {
+	name         string
+	copyTruncate bool
+}
+
+func (rt *rotationTest) Run(t *testing.T) {
+	t.Parallel()
+	f := NewFactory()
+	sink := &exportertest.SinkLogsExporter{}
+	params := component.ReceiverCreateParams{Logger: zaptest.NewLogger(t)}
+
+	tempDir := newTempDir(t)
+
+	// With a max of 100 logs per file and 1 backup file, rotation will occur
+	// when more than 100 logs are written, and deletion after 200 logs are written.
+	// Write 300 and validate that we got the all despite rotation and deletion.
+	logger := newRotatingLogger(t, tempDir, 100, 1, rt.copyTruncate)
+	numLogs := 300
+
+	cfg := f.CreateDefaultConfig().(*Config)
+	cfg.Pipeline = unmarshal(t, fmt.Sprintf(`
+- type: file_input
+  include: [%s/*]
+  start_at: beginning
+  poll_interval: 10ms`, tempDir))
+
+	rcvr, err := f.CreateLogsReceiver(context.Background(), params, cfg, sink)
+	require.NoError(t, err, "failed to create receiver")
+	require.NoError(t, rcvr.Start(context.Background(), &testHost{t: t}))
+
+	for i := 0; i < numLogs; i++ {
+		logger.Printf("This is a simple log line with the number %3d", i)
+		time.Sleep(time.Millisecond)
+	}
+
+	require.Eventually(t, expectNLogs(sink, numLogs), 2*time.Second, time.Millisecond)
+	require.NoError(t, rcvr.Shutdown(context.Background()))
+}
+
+func newRotatingLogger(t *testing.T, tempDir string, maxLines, maxBackups int, copyTruncate bool) *log.Logger {
+	path := filepath.Join(tempDir, "test.log")
+	rotator := &nanojack.Logger{
+		Filename:     path,
```

Comment:
Similar idea, but the extensions are timestamped with `2006-01-02T15-04-05.000000000` format, such as `test.log.2020-10-20T15-04-05.000000000`.

I can (and would like to) enhance `nanojack` to allow for various backup file naming conventions, but I will point out that what I think we're interested in here ultimately boils down to a question of whether or not `stanza` can correctly detect new files and files with previously seen names that have changed. I believe we cover these quite well in stanza unit tests, but I can appreciate that it would be worth covering here as well. I think enhancing `nanojack` to support `.n` extension should do both.

I will take a look at `nanojack` today.

Reply:
Tests have now been updated to include exactly this rotation pattern.

Reply:
Can you please add comments to explain this in the code?